File: 264_ugly_number_II.py
import heapq


class Solution:
    def nthUglyNumber(self, n: int) -> int:
        """
            Time Complexity - O(nlogn)
            Space Complexity - O(n)
        """
        pq = [1]
        visited = set([1])
        i = 1
        while i < n:
            cur_num = heapq.heappop(pq)
            for num in [2, 3, 5]:
                num *= cur_num
                if num not in visited:
                    heapq.heappush(pq, num)
                    visited.add(num)
            i += 1
        return heapq.heappop(pq)

    def nthUglyNumber(self, n: int) -> int:
        """
            Time Complexity - O(n)
            Space Complexity - O(1)
        """
        p2 = p3 = p5 = 0
        n2, n3, n5 = 2, 3, 5
        arr = [1]
        while len(arr) < n:
            cur_min = min(n2, n3, n5)
            arr.append(cur_min)
            if cur_min == n2:
                p2 += 1
                n2 = arr[p2] * 2
            if cur_min == n3:
                p3 += 1
                n3 = arr[p3] * 3
            if cur_min == n5:
                p5 += 1
                n5 = arr[p5] * 5
        return arr[n - 1]


if __name__ == '__main__':
    print(Solution().nthUglyNumber(10))

File: 264-UglyNumberII.cpp
/* 
    Time Complexity                              :  O(N)
    Space Complexity                             :  O(N)
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  


class Solution {
public:
    int nthUglyNumber(int n) {
        if(n == 1) return 1;
        
        int p2=0, p3=0, p5=0;
        vector<int> ans(n);
        ans[0]=1;
        int i=1;
        
        while(i<n) {
            
            int m2 = 2*ans[p2];
            int m3 = 3*ans[p3];
            int m5 = 5*ans[p5];
            
            int mn = min(m2, min(m3,m5));
            ans[i] = mn;
            
            if(m2 == mn)
                p2++;
            if(m3 == mn)
                p3++;
            if(m5 == mn)
                p5++;
            
            i++;
        }
        
        return ans[n-1];
    }
};